<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Function reference â€¢ svBreaks</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Function reference" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-index">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">svBreaks</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index">

    <colgroup>
      
      <col class="alias" />
      <col class="title" />
    </colgroup>

    <tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-all-functions" class="hasAnchor"><a href="#section-all-functions" class="anchor"></a>All functions</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="Volcano.html">Volcano()</a></code> </p>
        </td>
        <td><p>Plot the enrichment of SVs in genomic features</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="bpAllGenes.html">bpAllGenes()</a></code> </p>
        </td>
        <td><p>Calculate the enrichment of genes affected by SVs (as opposed to breakpoint hits)</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="bpFeatureEnrichment.html">bpFeatureEnrichment()</a></code> </p>
        </td>
        <td><p>Calculate the enrichment of SVs in genomic features</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="bpFeatureEnrichmentPlot.html">bpFeatureEnrichmentPlot()</a></code> </p>
        </td>
        <td><p>Plot the enrichment of SVs in genomic features</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="bpFeatures.html">bpFeatures()</a></code> </p>
        </td>
        <td><p>bpFeatures
Get some basic stats for breakpoints</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="bpGeneEnrichment.html">bpGeneEnrichment()</a></code> </p>
        </td>
        <td><p>Calculate the enrichment of SVs in genes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="bpGeneEnrichmentPlot.html">bpGeneEnrichmentPlot()</a></code> </p>
        </td>
        <td><p>Plot the enrichment of SVs in genes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="bpRainfall.html">bpRainfall()</a></code> </p>
        </td>
        <td><p>bpRainfall
Plot log10 distances between snvs as rainfall plot</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="bpRegionEnrichment.html">bpRegionEnrichment()</a></code> </p>
        </td>
        <td><p>Calculate the enrichment of SVs in genomic regions</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="bpRegionEnrichmentPlot.html">bpRegionEnrichmentPlot()</a></code> </p>
        </td>
        <td><p>Plot the enrichment of SVs in genomic features</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="bpRegioneR.html">bpRegioneR()</a></code> </p>
        </td>
        <td><p>Run regioneR for to regions in bed format</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="bpSim.html">bpSim()</a></code> </p>
        </td>
        <td><p>bpSim</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="bpStats.html">bpStats()</a></code> </p>
        </td>
        <td><p>bpStats
Get some basic stats for breakpoints</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="cleanTheme.html">cleanTheme()</a></code> </p>
        </td>
        <td><p>Set theme for clean plotting with ggplot2</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="dist2Motif.html">dist2Motif()</a></code> </p>
        </td>
        <td><p>dist2Motif
Calculate the distance from each breakpoint to closest motif</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="dist2motif2.html">dist2motif2()</a></code> </p>
        </td>
        <td><p>dist2Motif2
Calculate the distance from each breakpoint to closest motif in a directory of files</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="distOverlay.html">distOverlay()</a></code> </p>
        </td>
        <td><p>Calculate the distance from each breakpoint to closest motif
Overlay the same number of random simulated breakpoints</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="distOverlay2.html">distOverlay2()</a></code> </p>
        </td>
        <td><p>Calculate the distance from each breakpoint to closest motif
Overlay the same number of random simulated breakpoints</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="featureDensity.html">featureDensity()</a></code> </p>
        </td>
        <td><p>featureDensity</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="geneHit.html">geneHit()</a></code> </p>
        </td>
        <td><p>Print barplot showing samples with SVs affecting specified gene</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="generateData.html">generateData()</a></code> </p>
        </td>
        <td><p>generateData
Prepare data for dist2motif</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="generateData2.html">generateData2()</a></code> </p>
        </td>
        <td><p>generateData2
Prepare data for dist2motif</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="genesbySample.html">genesbySample()</a></code> </p>
        </td>
        <td><p>genesbySample
Plot the number of structural breakpoints per sample</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="genomeHits.html">genomeHits()</a></code> </p>
        </td>
        <td><p>genomeHits</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="genotypeTypeCount.html">genotypeTypeCount()</a></code> </p>
        </td>
        <td><p>genotypeTypeCount
Plot the contribution of sv tpyes to toal load per genotype</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getData.html">getData()</a></code> </p>
        </td>
        <td><p>getData</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getMissingSamples.html">getMissingSamples()</a></code> </p>
        </td>
        <td><p>get sample names</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="ggVolcano.html">ggVolcano()</a></code> </p>
        </td>
        <td><p>Plot the enrichment of SVs in genomic features</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="hello.html">hello()</a></code> </p>
        </td>
        <td><p>Hello, World!</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="mechansimSize.html">mechansimSize()</a></code> </p>
        </td>
        <td><p>Plot mechanims vs length
mechansimSize</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="mergeOverlaps.html">mergeOverlaps()</a></code> </p>
        </td>
        <td><p>Implements bedtools merge to merge overlapping regions in bedfile</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="micromologyPlot.html">micromologyPlot()</a></code> </p>
        </td>
        <td><p>Plot micromology vs type
micromology</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="notchFilt.html">notchFilt()</a></code> </p>
        </td>
        <td><p>Functions related to Notch
notchFilt</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="notchHits.html">notchHits()</a></code> </p>
        </td>
        <td><p>notchHits
Plot the breakpint density over Notch</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotMonteCarlo.html">plotMonteCarlo()</a></code> </p>
        </td>
        <td><p>Plot the result of a monte carlo simulation (n shuffles of feature y)
Expects dataframe returned from bpRegionEnrichment()</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotdistanceOverlay.html">plotdistanceOverlay()</a></code> </p>
        </td>
        <td><p>plotdistanceOverlay</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotdistanceOverlay2.html">plotdistanceOverlay2()</a></code> </p>
        </td>
        <td><p>plotdistanceOverlay</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="setCols.html">setCols()</a></code> </p>
        </td>
        <td><p>Set colour scheme for manual colouring</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="showSamples.html">showSamples()</a></code> </p>
        </td>
        <td><p>showSamples</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="simSig.html">simSig()</a></code> </p>
        </td>
        <td><p>simSig</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="simSig2.html">simSig2()</a></code> </p>
        </td>
        <td><p>Calculate stats for simulated data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sizeDist.html">sizeDist()</a></code> </p>
        </td>
        <td><p>sizeDist
Plot the size distribution for different types of structual variants</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="slideTheme.html">slideTheme()</a></code> </p>
        </td>
        <td><p>Set theme for slide plotting with ggplot2</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="snv_colours.html">snv_colours()</a></code> </p>
        </td>
        <td><p>SNV colours</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="subtractUnmappable.html">subtractUnmappable()</a></code> </p>
        </td>
        <td><p>Implements bedtools subtract to remove unmappable regions in bedfiles</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="svTypes.html">svTypes()</a></code> </p>
        </td>
        <td><p>svTypes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sv_colours.html">sv_colours()</a></code> </p>
        </td>
        <td><p>SV colours</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="svsbySample.html">svsbySample()</a></code> </p>
        </td>
        <td><p>svsbySample
Plot the number of structural breakpoints per sample</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="swapSampleNames.html">swapSampleNames()</a></code> </p>
        </td>
        <td><p>Swap sample names</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="tally_hits.html">tally_hits()</a></code> </p>
        </td>
        <td><p>Tally type of sv events in specified gene</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="transform_types.html">transform_types()</a></code> </p>
        </td>
        <td><p>Misc functions
transform_types</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="writeBed.html">writeBed()</a></code> </p>
        </td>
        <td><p>Simple function to write out bedfiels - useful for checking after merging/subtracting...</p></td>
      </tr>
    </tbody>
    </table>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Nick Riddiford.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


